notebooks/customizing_potentials_cookbook.ipynb

```diff
@@ -0,0 +1,1780 @@
+        "\n",
+        "**Cell lists**\n",
+        "\n",
+        "The technique here is to divide space into small cells that are just larger than the largest interaction range in the system. Thus, if particle $i$ is in cell $c_i$ and particle $j$ is in cell $c_j$, $i$ and $j$ can only interact if $c_i$ and $c_j$ are neighboring cells. Rather than searching all $N^2$ combinations of particle pairs for non-zero interactions, you only have to search the particles in the neighboring cells. \n",
+        "\n",
+        "**Neighbor lists**\n",
+        "\n",
+        "The technique here is, for each particle $i$, to make a list of *potential* neighbors: particles $j$ that are within some threshold distance $r_\\mathrm{cutoff}$. If $r_\\mathrm{cutoff} = r_\\mathrm{max} + r_\\mathrm{buffer}$ (where $r_\\mathrm{max}$ is the largest interaction range in the system and $r_\\mathrm{buffer}$ is an appropriately chosen buffer size), then all interacting particles will appear in this list as long as no particles moves by more than $r_\\mathrm{buffer}/2$. There is a tradeoff here: smaller $r_\\mathrm{buffer}$ means fewer particles to search over each MD step but the list must be recalculated more often, while larger $r_\\mathrm{buffer}$ means slower force calculates but less frequent neighbor list calculations. \n",
+        "\n",
+        "In practice, the most efficient technique is often to use cell lists to calculate neighbor lists. In JAX MD, this occurs under the hood, and so only calls to neighbor-list functionality are necessary."
+      ]
+    },
+    {
+      "cell_type": "code",
+      "metadata": {
+        "id": "9QXA1k8oQcwk",
+        "colab_type": "code",
+        "colab": {}
+      },
+      "source": [
+        "def harmonic_morse_cutoff_neighbor_list(\n",
+        "    displacement_or_metric,\n",
+        "    box_size,\n",
+        "    example_R,\n",
+        "    species=None,\n",
+        "    D0=5.0, \n",
+        "    alpha=5.0, \n",
+        "    r0=1.0, \n",
+        "    k=50.0,\n",
+        "    r_onset=1.0,\n",
+        "    r_cutoff=1.5, \n",
+        "    list_cutoff=2.0, \n",
+        "    buffer_size_multiplier=3.0,\n",
+        "    **kwargs): \n",
+        "\n",
+        "  D0 = np.array(D0, dtype=np.float32)\n",
+        "  alpha = np.array(alpha, dtype=np.float32)\n",
+        "  r0 = np.array(r0, dtype=np.float32)\n",
+        "  k = np.array(k, dtype=np.float32)\n",
+        "  r_onset = np.array(r_onset, dtype=np.float32)\n",
+        "  r_cutoff = np.array(r_cutoff, np.float32)\n",
+        "  list_cutoff = np.float32(list_cutoff)\n",
+        "\n",
+        "  neighbor_fn = partition.neighbor_list(\n",
+        "    displacement_or_metric, \n",
+        "    box_size, \n",
+        "    list_cutoff, \n",
+        "    example_R, \n",
+        "    buffer_size_multiplier=buffer_size_multiplier)\n",
+        "  \n",
+        "  energy_fn = smap.pair_neighbor_list(\n",
```

Comment:
That's not a bad idea, and would definitely make it easier to switch back and forth. Probably pretty trivial on your end to make the change? 

One other related question: is it possible to use cell lists directly without neighbor lists? My guess is that in most cases this would be sub-optimal, but they have the benefit of having zero free parameters (e.g. like ```dr_threshold```)... not sure if the benefit is worth the effort.

Another though about cell lists while we are on the subject... I wonder how hard it would be to implement the version of cell lists where you make the cells smaller than the max interaction range but then loop over more cells (most of which are empty). The idea is that if you can guarantee that particles will never be _closer_ than some r_min, then you can set the cells so that the max occupancy is guaranteed to be 1 for all cells. I genuinely have no idea if this would be an improvement...

Reply:
One other comment about cell lists without neighbor lists. And this is part of a bigger discussion with a lot of possible solutions that I might not have thought about. But neighbor lists seem to have a hard time with batching / ensembling because `nbrs = neighbor_fn(R, nbrs)` doesn't play well with vmap. 

I've been playing around with potential work-arounds to this, but they seem fairly slow / unsafe, though its definitely possible (likely?) that there's a simple solution I don't see. But if I understand it right, I _think_ cell lists could work well with vmap since every system does the exact same calculation every step.